pennylane/transforms/decompose.py

```diff
@@ -0,0 +1,157 @@
+            raise UserWarning(
+                f"Operator {op.name} has no supported decomposition and is not in the gate set."
+            ) from e
+
+        for sub_op in decomp:
+            yield from _operator_decomposition_gen(
+                sub_op,
+                acceptance_function,
+                decomposer=decomposer,
+                max_expansion=max_expansion,
+                current_depth=current_depth,
+            )
+
+
+@transform
+def apply_decomposition(tape, gate_set=None, gate_rules=None, max_expansion=None):
+    """Decomposes quantum circuit into a desired gate set.
+
+    Args:
+        tape (QuantumScript or QNode or Callable): a quantum circuit.
+        gate_set (set, optional): A set of decomposition gates. Defaults to None. If ``None``, gate set defaults to all operators not currently present.
+        gate_rules (Callable[qml.operation.Operator, bool], optional): A rule set for the gate set to follow. Defaults to None.
+        max_expansion (int, optional): The maximum depth of the expansion. Defaults to None.
+
+    Returns:
+        qnode (QNode) or quantum function (Callable) or tuple[List[QuantumScript], function]:
+
+        The decomposed circuit. The output type is explained in :func:`qml.transform <pennylane.transform>`.
+
+    .. seealso:: :func:`~.pennylane.devices.preprocess.decompose` for a transform that is designed particularly for restricted gate sets on specific device architectures.
+
+    **Examples:**
+
+    >>> tape = qml.tape.QuantumScript([qml.IsingXX(1.2, wires=(0,1))], [qml.expval(qml.Z(0))])
+    >>> batch, fn = apply_decomposition(tape, gate_set = {"CNOT", "RX", "RZ"})
+    >>> batch[0].circuit
+    [CNOT(wires=[0, 1]),
+    RX(1.2, wires=[0]),
+    CNOT(wires=[0, 1]),
+    expval(Z(0))]
+
+    >>> @partial(apply_decomposition, gate_rules = lambda obj: len(obj.wires) <= 2)
+    >>> @qml.qnode(device)
+    >>> def circuit():
+    >>>     qml.Toffoli(wires = range(3))
+    >>>
+    >>> print(qml.draw(circuit)())
+    0: ───────────╭●───────────╭●────╭●──T──╭●─┤
+    1: ────╭●─────│─────╭●─────│───T─╰X──T†─╰X─┤
+    2: ──H─╰X──T†─╰X──T─╰X──T†─╰X──T──H────────┤
+
+    """
+    universal_ops = set(qml.ops.__all__)
+
+    if gate_set is None:
+        gate_set = universal_ops - set(op.name for op in tape.operations)
+
+    if gate_rules is None:
+
+        def gate_rules(op):
+            return not op.has_matrix
+
+    if isinstance(gate_set, str):
+        gate_set = set([gate_set])
+
+    if isinstance(gate_set, (list, tuple)):
+        gate_set = set(gate_set)
+
+    def decomposer(op):
+        return op.decomposition()
+
+    def stopping_condition(op):
+        if not op.has_decomposition:
+            if op.name not in gate_set:
+                warnings.warn(
+                    f"Operator {op.name} has no supported decomposition and was not found in the gate set.",
+                    UserWarning,
+                )
+            return True
```

Comment:
Sorry, I must not have read through carefully enough before! I just realized you are catching everything that doesn't have a decomposition here anyway, so that it passes the acceptance criteria 😅  

In that case there's some redundancy the `_operator_decomposition_gen` above that could be removed completely, instead of updated to give a warning like I suggested earlier.